File: tools/session-state/src/session_state/keying.py
"""keying.py — repo→state-dir keying.

canonical_project_root / path_to_slug are VENDORED from
~/.claude/tools/context-offload/recall.py (do not import across tools — separate install
roots). test_keying.py asserts parity. If recall.py changes, re-vendor.
"""
from __future__ import annotations

import json
import os
import subprocess
from pathlib import Path


def claude_home() -> Path:
    override = os.environ.get("CLAUDE_CONFIG_DIR")
    return Path(override) if override else Path.home() / ".claude"


def canonical_project_root(cwd: Path) -> Path:
    """Collapse `<root>/.claude/worktrees/<name>` back to `<root>` (vendored from recall.py)."""
    parts = cwd.parts
    for i in range(len(parts) - 1):
        if parts[i] == ".claude" and i + 1 < len(parts) and parts[i + 1] == "worktrees":
            return Path(*parts[:i])
    return cwd


def path_to_slug(path: Path) -> str:
    """Claude Code's project slug: replace separators and drive colon with '-' (vendored)."""
    s = str(path)
    return s.replace("\\", "-").replace("/", "-").replace(":", "-")


def repo_root(cwd: Path) -> Path | None:
    try:
        out = subprocess.run(
            ["git", "-C", str(cwd), "rev-parse", "--show-toplevel"],
            capture_output=True, text=True, timeout=5,
        )
        if out.returncode != 0:
            return None
        return canonical_project_root(Path(out.stdout.strip()))
    except Exception:
        return None


def repo_key(cwd: Path) -> str:
    root = repo_root(cwd)
    return path_to_slug(root if root is not None else cwd)


def state_dir(cwd: Path) -> Path:
    d = claude_home() / "state" / "sessions" / repo_key(cwd)
    (d / "scratch").mkdir(parents=True, exist_ok=True)
    return d


def check_meta(dir: Path, canonical_repo: str) -> bool:
    """Write meta.json on first use; return False (and drop a COLLISION marker) on mismatch."""
    meta = dir / "meta.json"
    if not meta.exists():
        meta.write_text(json.dumps({"repo": canonical_repo}), encoding="utf-8")
        return True
    try:
        existing = json.loads(meta.read_text(encoding="utf-8")).get("repo")
    except Exception:
        existing = None
    if existing == canonical_repo:
        return True
    (dir / "COLLISION").write_text(
        f"expected {existing!r} got {canonical_repo!r}", encoding="utf-8"
    )
    return False

File: plugins/session-state/src/keying.mjs
import { homedir } from "node:os";
import path from "node:path";
import fs from "node:fs";
import { execFileSync } from "node:child_process";

export function claudeHome() {
  return process.env.CLAUDE_CONFIG_DIR || path.join(homedir(), ".claude");
}

// Collapse "<root>/.claude/worktrees/<name>" back to "<root>" (string-based; sep-agnostic).
export function canonicalProjectRoot(cwd) {
  const m = cwd.match(/^(.*)[\\/]\.claude[\\/]worktrees[\\/][^\\/]+(?:[\\/].*)?$/);
  return m ? m[1] : cwd;
}

export function pathToSlug(p) {
  return p.replace(/[\\/:]/g, "-");
}

// Canonicalize separators for repo-path comparison/keying only (display is unaffected):
// "\" and "/" denote the same repo, so collapse to a single canonical separator before
// any compare. Applied to BOTH sides so a sep difference can never produce a mismatch.
function normSep(p) {
  return p == null ? p : p.replace(/\\/g, "/");
}

export function repoRoot(cwd) {
  try {
    const out = execFileSync("git", ["-C", cwd, "rev-parse", "--show-toplevel"],
      { encoding: "utf8", timeout: 5000 }).trim();
    return out ? canonicalProjectRoot(out) : null;
  } catch { return null; }
}

export function repoKey(cwd) {
  return pathToSlug(repoRoot(cwd) || cwd);
}

export function stateDir(cwd, { create = true } = {}) {
  const d = path.join(claudeHome(), "state", "sessions", repoKey(cwd));
  if (create) fs.mkdirSync(path.join(d, "scratch"), { recursive: true });
  return d;
}

export function checkMeta(dir, canonicalRepo) {
  const meta = path.join(dir, "meta.json");
  if (!fs.existsSync(meta)) {
    fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(meta, JSON.stringify({ repo: canonicalRepo }));
    return true;
  }
  let existing = null;
  let parsed = false;
  try {
    existing = JSON.parse(fs.readFileSync(meta, "utf8")).repo;
    parsed = true;
  } catch {}
  // Corrupt/unparseable meta, or a parsed meta with no repo recorded, is recoverable:
  // heal it by rewriting with the current repo. Only a parsed meta that records a
  // genuinely different repo is a real COLLISION.
  if (!parsed || existing == null) {
    fs.writeFileSync(meta, JSON.stringify({ repo: canonicalRepo }));
    return true;
  }
  // Compare separator-normalized: a "\" vs "/" difference is the same repo, not a collision.
  if (normSep(existing) === normSep(canonicalRepo)) return true;
  fs.writeFileSync(path.join(dir, "COLLISION"), `expected ${existing} got ${canonicalRepo}`);
  return false;
}
